src/Compilers/Core/CodeAnalysisTest/DefaultAnalyzerAssemblyLoaderTests.cs

```diff
@@ -37,54 +37,95 @@ public class AssemblyLoadTestFixtureCollection : ICollectionFixture<AssemblyLoad
+            //
+            // Not only is this bad behavior it also pollutes future test results.
+            var count = AssemblyLoadContext.Default.Assemblies.Count();
+            try
+            {
+                using var fixture = new AssemblyLoadTestFixture();
+                var loader = Create(alc);
+                DefaultAnalyzerAssemblyLoaderTestsBase.InvokeTestCode(loader, fixture, typeName, methodName);
+            }
+            finally
+            {
+                Assert.Equal(count, AssemblyLoadContext.Default.Assemblies.Count());
+            }
         }
+    }
+
+    public sealed class ShadowInvokeUtil : InvokeUtil
+    {
+        internal override DefaultAnalyzerAssemblyLoader Create(AssemblyLoadContext alc) =>
+            new ShadowCopyAnalyzerAssemblyLoader(compilerLoadContext: alc);
+    }
+
 #else
-        private sealed class InvokeUtil : MarshalByRefObject
+
+    public class InvokeUtil : MarshalByRefObject
+    {
+        internal virtual DefaultAnalyzerAssemblyLoader Create() =>
+            new DefaultAnalyzerAssemblyLoader();
+
+        public void Exec(string typeName, string methodName)
         {
-            public void Exec(string typeName, string methodName)
+            try
             {
-                try
-                {
-                    InvokeTestCode(typeName, methodName);
-                }
-                catch (TargetInvocationException ex) when (ex.InnerException is XunitException)
-                {
-                    var inner = ex.InnerException;
-                    throw new Exception(inner.Message + inner.StackTrace);
-                }
+                var loader = Create();
+                using var fixture = new AssemblyLoadTestFixture();
+                DefaultAnalyzerAssemblyLoaderTestsBase.InvokeTestCode(loader, fixture, typeName, methodName);
+            }
+            catch (TargetInvocationException ex) when (ex.InnerException is XunitException)
+            {
+                var inner = ex.InnerException;
+                throw new Exception(inner.Message + inner.StackTrace);
             }
         }
+    }
+
+    public sealed class ShadowInvokeUtil : InvokeUtil
+    {
+        internal override DefaultAnalyzerAssemblyLoader Create() =>
+            new ShadowCopyAnalyzerAssemblyLoader();
+    }
 #endif
 
-        private static readonly CSharpCompilationOptions s_dllWithMaxWarningLevel = new(OutputKind.DynamicallyLinkedLibrary, warningLevel: CodeAnalysis.Diagnostic.MaxWarningLevel);
-        private readonly ITestOutputHelper _output;
+    public class DefaultAnalyzerAssemblyLoaderTests : DefaultAnalyzerAssemblyLoaderTestsBase<InvokeUtil>
+    {
+    }
 
-        public DefaultAnalyzerAssemblyLoaderTests(ITestOutputHelper output)
-        {
-            _output = output;
-        }
+    public class ShadowCopyDefaultAnalyzerAssemblyLoaderTests : DefaultAnalyzerAssemblyLoaderTestsBase<ShadowInvokeUtil>
```

Comment:
I'm concerned this will negatively impact the dev experience when using the "run tests in context" command. If I go to a single test method and "run tests in context", the test framework will no longer have enough information to run the test.

Reply:
Correct. Once you move to an `abstract` type with derived types as the mechanism for providing behavior you lose the ability to right click and run test in the editor. You instead have to go through Test Explorer and run the test from there. 

I wish the tests had a better UI for this but alas they do not. There are several other examples of this pattern in our code base. It's workable. If there are suggestions for better way to plumb this through Im' interested in discussing.

I have thought about long term doing a runner extensibility point similar to `[WpfFact]` called `[AssemblyContextFact]` which effectively runs your test in a new assembly context. For .NET Core it's a new `AssemblyLoadContext` and for .NET framework it's `AppDomain`. That is a *lot* of work though and haven't gone that far yet.

Reply:
A few alternatives on my mind.

1. Make it so when individual tests call Run(), that we run the test code using both the regular and shadow copy loaders.
2. Make all the tests Theory, with the loader as a new test method parameter.

If it feels like there are other considerable benefits of the current pattern which outweigh the above alternatives, then I won't *die on that hill*, but it means I will be fighting my muscle memory whenever I come back here in the future 😉

Reply:
I _think_ that either is doable. Essentially what would happen is either:

1. The implementation of `Run` calls the delegate twice, once with default and the other with shadow copy loader 
2. Have a `bool isDefault` parameter to the test case which feeds into `Run` and the implementation then toggles if it's a shadow copy or not. 

I'm not sure which is the better approach.

Reply:
After some thought I'm going with 2. The rational is I feel like it's easier to debug. 

Consider how failures manifest in CI logs. Unless we are very careful with 1 it will be very difficult to tell if the failure was in the default load vs shadow load case. Nothing on the stack would give you a clue that was the case. Where as with 2 it's a part of the test name so it's a lot more obvious.

Reply:
Okay I'm going with 2. It was relatively straight forward to implement it. I've had to track down several failures locally and CI and I'm satisfied the information is there to make it clear which context the failure occurred in.